metabase/metabase
# Migrate field ref-based column keys in `column_settings`

Related to https://github.com/metabase/metabase/pull/46383

We had a lot of migrations to maintain field refs in `column_settings` in the past, and I hope this is the last migration.

Currently, we have 2 column keys used in `column_settings` - field ref-based (`["ref", ["field", 1, null]]`) and name-based (`["name", "ID"]`). We use(d) field-ref keys for simple cases where a query is based on a raw table, and name-based in more complex cases (this is simplified). This was error prone because of the following reasons:
- Makes it very hard to do any MBQL changes/fixes because every change would require a custom migration
- Field ref-based keys omit `:binning` and `temporal-unit` options, making it impossible to have columns that differ only by this options
- Some changes to a query (new stage etc) would trigger field ref changes and loss of all settings
- Any bugs with field refs (self joins etc) automatically result in bugs in viz settings

This PR migrates all `column_settings` keys to name-based keys. The FE in the referenced PR supports both field ref- and name-based keys for the keys that would not be migrated by this migration (no `result_metadata`).

How to verify:
- Switch to the base branch (master)
- Create questions with `column_settings`. These are formatting settings for columns. E.g. New -> Orders -> Visualize, click on column headers -> Gear icon -> Change column title -> Save question
- Checkout this branch and run
- All your viz settings should be migrated. There should be no visual changes. Export should continue working and it should use the updated settings.
- You can check that the settings are migrated by inspecting Network and making sure that `["name", ...]` keys are used in `column_settings` in `visualization_settings`.
